src/modules/Block.ts

```diff
@@ -0,0 +1,201 @@
+    this._element.insertAdjacentElement('afterbegin', element);
+    this._removeEvents();
+    this._addEvents();
+  }
+
+  _preCompileTemplate() {
+    return Object.entries(this.props)
+      .reduce((acc, item) => {
+        const [ key, instance ] = item;
+        const isBlockArray = Array.isArray(instance) && instance.every(item => item instanceof Block);
+        if (instance instanceof Block || isBlockArray) {
+          return { ...acc, [key]: makeComponentContainer(key) };
+        }
+        return { ...acc, [key]: instance };
+      }, {});
+  }
+
+  _removeEvents(): void {
+		const { events = {} } = this.props;
+		Object.keys(events).forEach((eventName) => {
+			this._element.removeEventListener(eventName, events[eventName]) ;
+		});
+	}
+
+  _addEvents(): void {
+		const { events = {} } = this.props;
+		Object.keys(events).forEach(eventName => {
+			this._element.addEventListener(eventName, events[eventName]);
+		});
+	}
+
+  _compileDomElement(tmpl: string): HTMLElement {
+    const { body } = new DOMParser().parseFromString(tmpl, 'text/html');
+    const element = body?.children[0] as HTMLElement;
+
+    if (this.props && !!element) {
+      Object.entries(this.props)
+        .forEach((item) => {
+          const [ key, instance ] = item;
+          const subComponents = element.querySelectorAll(`[data-component="${key}"]`);
+          const subComponentsParent = subComponents[0]?.parentElement;
+
+          if (subComponents.length === 0) {
+            return element;
+          }
+
+          if (Array.isArray(instance)) {
+            instance.forEach((instanceEl: Block) => subComponents.forEach(subComponent => {
+              subComponent.insertAdjacentElement('beforebegin', instanceEl.getContent())
+            }
+            ));
+          } else {
+            subComponents.forEach(subComponent => subComponent.insertAdjacentElement('beforebegin', instance.getContent()))
+          }
+
+          subComponents.forEach(subComponent => subComponentsParent.removeChild(subComponent));
+      })
+    }
+    return element;
+  }
+
+  get element(): HTMLElement {
+		return this._element;
+	}
+
+  getContent(): HTMLElement {
+		return this.element;
+	}
+
+  render(): Function {
+    return () => {};
+  }
+
+  isPropsChanged<T>(oldProps: T, newProps: T): boolean {
+    // TODO: add shallow equal
+    return true;
+  }
+
+  setProps(nextProps: Props): void {
+		if (!nextProps) {
+			return;
+		}
+
+		Object.assign(this.props, nextProps);
+	};
+
+  _makePropsProxy(props: Props): Props {
+    const self = this;
+    const oldProps: Props = JSON.parse(JSON.stringify(props));
+
+    return new Proxy(props, {
+      get(target: Props, prop: string) {
+        const value = target[prop];
+        return typeof value === 'function' ? value.bind(target) : value;
+      },
+      set(target: Props, prop: string, value): boolean {
+        target[prop] = value;
+        self._componentDidUpdate(oldProps, target);
```

Comment:
Можно использовать стрелочные функции, тогда не придется прибегать к замыканию
```suggestion
      set: (target: Props, prop: string, value): boolean => {
        target[prop] = value;
        this._componentDidUpdate(oldProps, target);
```